Stat testing (#27)
tidy-survey-r/tidy-survey-book: 07-statistical-testing.Rmd

# Statistical testing {#c07}

When analyzing results from a survey, the point estimates described in Chapter \@ref(c06) are helpful for understanding the data at a high level, but researchers and the public often want to make comparisons between different groups. These comparisons are calculated through statistical testing where we compare the point estimates and the variance estimates of each statistic to see if there are statistically significant differences. The general idea of statistical testing is the same for data obtained through surveys and data obtained through other methods, but the importance lies in ensuring the variance is calculated correctly. Functions in the {survey} packages allow for the correct estimation of the variances. This chapter will cover the following statistical tests with survey data and functions:

-  Comparison of proportions `svyttest()`
-  Comparison of means `svyttest()`
-  Goodness of fit tests `svygofchisq()`
-  Tests of independence `svychisq()`
-  Tests of homogeneity `svychisq()`

> Up to this point, the functions that we've provided have used the wrappers from the {srvyr} package. This means that the functions work with tidyverse syntax. However, the functions in this chapter do not have wrappers from the {srvyr} package and are instead pulled directly from the {survey} package. This means that the design object is *not* the first argument and that to use these functions with the magrittr pipe `%>%` and tidyverse syntax we will need to use dot (`.`) notation. Functions that work with the magrittr pipe `%>%` have the first argument as data. When data is run, it automatically places anything to the left of the pipe into the first argument of the function to the right of the pipe. For example, if we wanted to take the `mtcars` data and filter to only cars with 6 cylinders we can write the code in one of three ways:

Reviewer: ```suggestion
Up to this point, the functions that we've provided have used the wrappers from the {srvyr} package. This means that the functions work with tidyverse syntax. However, the functions in this chapter do not have wrappers from the {srvyr} package and are instead used directly from the {survey} package. This means that the design object is *not* the first argument and that to use these functions with the magrittr pipe `%>%` and tidyverse syntax we will need to use dot (`.`) notation. Functions that work with the magrittr pipe `%>%` have the first argument as data. When data is run, it automatically places anything to the left of the pipe into the first argument of the function to the right of the pipe. For example, if we wanted to take the `mtcars` data and filter to only cars with 6 cylinders we can write the code in one of three ways:
```

1. `filter(mtcars, cyl == 6)`
2. `mtcars %>% filter(cyl == 6)`

> 3. `mtcars %>% filter(., cyl == 6)`

Reviewer: Make it explicit that the argument is .data. Useful for later, I think

```suggestion
3. `mtcars %>% filter(.data = ., cyl == 6)`
```

> 3. `mtcars %>% filter(., cyl == 6)`
> 

Reviewer: ```suggestion
3. `mtcars %>% filter(., cyl == 6)`
4. `mtcars %>% filter(.data= ., cyl == 6)`
```

> Each of these lines of code will produce the same output since the argument that takes the data is in the first spot in `filter()`. Those who have worked with the tidyverse, the first two are probably familiar. The third option functions the same way as the second one, but is explicit about where the data is going. Here we're telling R to take what's on the left side of the pipe `mtcars` and pipe it into the spot with the dot (`.`)---the first argument.

Reviewer: A bit more to talk about the named argument option as well

```suggestion
Each of these lines of code will produce the same output since the argument that takes the data is in the first spot in `filter()`. Those who have worked with the tidyverse, the first two are probably familiar. The third option functions the same way as the second one, but is explicit about where that `mtcars` goes in the first argument and the fourth option indicates that mtcars is going into the named argument of `.data`. Here we're telling R to take what's on the left side of the pipe `mtcars` and pipe it into the spot with the dot (`.`)---the first argument.
```

In functions that are not part of the tidyverse, the data argument may be in a different spot in the functions. For example, in `svyttest()` the data argument is in the second spot, which means we need to place the dot (`.`) in the second spot and not the first. For example:

> ```{r stattest-dotex}
> #| label: stattest-dotex
> #| eval: FALSE
> svydata_des %>% 
>   svyttest(x~y,.)
> ```

Reviewer: Change to code chunk, makes easier to run interactive with (run all above)

````suggestion
```
#| label: stattest-dotex

svydata_des %>% 
  svyttest(x~y,.)
```
````

Placing the dot (`.`) in the second argument spot, indicates that the survey design object `svydata_svy` should be used in the second argument and not the first (which is the default). 

> 

Reviewer: Alternatively, named arguments could be used to place the dot first as in the following:

```
svydata_des %>% 
  svyttest(design = ., x~y)
```

> 

Reviewer: Add another suggestion with design coming first. I like this a bit better because it's similar to how {srvyr} handles it.

````suggestion
Alternatively, named arguments could be used to place the dot first as in the following:

```
svydata_des %>% 
  svyttest(design = ., x~y)
```
````

## Chapter Set-Up {#stattest-setup}

For this chapter, we will be using the same data as we did in \@ref(c06): ANES and RECS. As a reminder, we will need to create survey design objects to work with. These design objects ensure that the variance estimation is calculated accurately, and thus we can accurately determine statistical significance.

First, make sure install and load the following packages:
```{r stattest-pkgs}
#| error: FALSE
#| warning: FALSE
#| message: FALSE

library(tidyverse)
library(survey) # for survey analysis
library(srvyr) # for tidy survey analysis
library(readr)
library(here)
library(gt) # for output of readable tables
```

Second, we need to read in the data and create the design objects.

Here is how to create the design object for the ANES data, remember that we need to adjust the weight so it sums to the population instead of the sample. We do that by multiplying the weights (see the ANES methodology documentation for more information).
```{r stattest-anesdes}

> anes <- read_rds(here("data", "anes_2020.rds")) %>%

Reviewer: Update file location 

```suggestion
anes <- read_rds(here("AnalysisData", "anes_2020.rds")) %>%
```

  mutate(Weight=Weight/sum(Weight)*231592693) 

anes_des <- anes %>%
  as_survey_design(weights = Weight,
          strata = Stratum,
          ids = VarUnit,
          nest = TRUE)
```

Here is how to create the design object for the RECS data:
```{r stattest-recsdes}

> recs <-read_rds(here("data", "recs.rds"))

Reviewer: Update location/file name

```suggestion
recs <-read_rds(here("AnalysisData", "recs_2015.rds"))
```

The difference between using t-tests with non-survey data and with survey data is based on the underlying variance estimation difference. Chapter \@ref(c05) provides the detailed overview of the math behind the mean and sampling error calculations for various sample designs. The functions in the {survey} package will account for these nuances provided the design object is correctly defined.  

### Syntax {#stattest-ttest-syntax}

When we do not have survey data, we may be able to use the `t.test()` function. This function does not allow for weights or the variance structure to be accounted for with survey data. Therefore, when using survey data, we need to use the `svyttest()` function. Many of the arguments are the same between the two functions, but there are a two key differences:

- We need to use the survey design object, instead of data
- We can only use a formula and not separate x and y data

Here is the syntax for the `svyttest()` function:

> ``` {r}
> #| label: stattest-ttest-syntax
> #| eval: FALSE
> svyttest(formula, 
>          design, 
>          na.rm=FALSE,
>          level=0.95,
>          ...)
> ```

Reviewer: ````suggestion
```
svyttest(formula, 
         design, 
         na.rm=FALSE,
         level=0.95,
         ...)
```
````

Notice that the first argument here is the `formula` and not the `design`. This means that we must use the dot `(.)` if we pipe in the survey design object (as described at the beginning of this chapter). 

The `formula` argument can take on a couple of different forms depending on what it is that we are measuring.  Here are a few common scenarios:

> 1. **One-sided t-test:** 

Reviewer: When I think of sided tests, I think of two sided having an alternative hypothesis of not equal but a one-sided would have an alternative hypothesis of > or <. This language makes it confusing to other texts.

Author: Good catch!

    a. **Comparison to 0:** $var \sim 0$, where $var$ is the measure of interest and $0$ is the value we are comparing it to. For example, we could test if the proportion of the population that has blue eyes is different from $0$.
    b. **Comparison to a different value:** $I(var - value) \sim 0$, where $var$ is the measure of interest and $value$ is what we are comparing to. We need to use the `I()` function, to tell the program to calculate the difference between the variable and the comparison value prior to testing. For example, we could test if the proportion of the population that has blue eyes is different from $25%$.
2. **Two-sided t-test:**
    a. **Unpaired:**

>         - **2 level grouping variable:** $var \sim groupVar$, where $var$ is the measure of interest and $groupVar$ is a variable with two categories.  For example, we could test if the perfect of the popualtion that has blue eyes is different for men and women. 

Reviewer: Note here or elsewhere: The group variable must be a factor or character with two levels, or be coded 0/1 or 1/2

> 1. **One-sided t-test:** 
>     a. **Comparison to 0:** $var \sim 0$, where $var$ is the measure of interest and $0$ is the value we are comparing it to. For example, we could test if the proportion of the population that has blue eyes is different from $0$.
>     b. **Comparison to a different value:** $I(var - value) \sim 0$, where $var$ is the measure of interest and $value$ is what we are comparing to. We need to use the `I()` function, to tell the program to calculate the difference between the variable and the comparison value prior to testing. For example, we could test if the proportion of the population that has blue eyes is different from $25%$.
> 2. **Two-sided t-test:**
>     a. **Unpaired:**
>         - **2 level grouping variable:** $var \sim groupVar$, where $var$ is the measure of interest and $groupVar$ is a variable with two categories.  For example, we could test if the perfect of the popualtion that has blue eyes is different for men and women. 
>         - **3+ level grouping variable:** $var \sim I(groupVar == level)$, where $var$ is the measure of interest, $groupVar$ is the categorical variable, and $level$ is the category level to isolate. Again we need to use the `I()` function to tell the program to isolate the category before doing the comparison across groups.  For example, we could test if the test scores in one classroom differed from all other classrooms.
>     b. **Paired:** $I(var_1 - var_2) \sim 0$, where $var_1$ is the first variable of interest and $var_2$ is the second variable of interest. We again will have to use the `I()` function to have the program calculate the difference between the two variables before comparing it against $0$.  For example, we could test if test scores on a subject differed between the start and the end of a course.

Reviewer: I think we need code syntax and not equation. Also might want to get rid of an example with a gender binary.

```suggestion
1. **One-sided t-test:** 
    a. **Comparison to 0:** `var ~ 0`, where `var` is the measure of interest and `0` is the value we are comparing it to. For example, we could test if the proportion of the population that has blue eyes is different from `0`.
    b. **Comparison to a different value:** `I(var - value) ~ 0`, where `var` is the measure of interest and `value` is what we are comparing to. We need to use the `I()` function, to tell the program to calculate the difference between the variable and the comparison value prior to testing. For example, we could test if the proportion of the population that has blue eyes is different from `25%` by using `I(var - 0.25)~0`.
2. **Two-sided t-test:**
    a. **Unpaired:**
        - **2 level grouping variable:** `var ~ groupVar`, where `var` is the measure of interest and `groupVar` is a variable with two categories.  For example, we could test if the proportion of the population that has blue eyes is different for men and women. 
        - **3+ level grouping variable:** `var ~ I(groupVar == level)`, where `var` is the measure of interest, `groupVar` is the categorical variable, and `level` is the category level to isolate. Again we need to use the `I()` function to tell the program to isolate the category before doing the comparison across groups.  For example, we could test if the test scores in one classroom differed from all other classrooms.
    b. **Paired:** `I(var_1 - var_2) ~ 0`, where `var_1` is the first variable of interest and `var_2` is the second variable of interest. We again will have to use the `I()` function to have the program calculate the difference between the two variables before comparing it against `0`.  For example, we could test if test scores on a subject differed between the start and the end of a course.
```

Additionally, the `na.rm` argument defaults to `FALSE`, which means if any data is missing the t-test will not compute. Throughout this chapter we will always set `na.rm=TRUE`, but before analyzing the survey data, make sure to review the notes provided in Chapter \@ref(c03) to better understand how to handle missing data.  Finally, the `level` argument is $1-\alpha$, or the amount of type 1 error.  The default is $0.95$.

### Examples {#stattest-ttest-examples}
Let's walk through a few examples using the ANES and RECS data.  See Section \@ref(stattest-setup) above to set up the design objects.

#### Example 1: One-sided t-test {.unnumbered #stattest-ttest-ex1}

ANES asked respondents if they voted early in the 2020 election.  In our data, we've called this variable `EarlyVote2020`.  If we want to see if the proportion of the U.S. voting eligible population that early voted is greater than 0%, we could set up the hypothesis as follows:

- $H_0: p = 0$ where $p$ is the proportion of U.S. voting eligible population that voted early

> - $H_A: p \neq 0$

Reviewer: Not loving this example. Maybe anchor it to something other than 0%. Something like, in xyz country, 75% of people vote in national elections. Do US citizens vote at the same rate?

To conduct this in R, we would then use the `svyttest()` function:
```{r}
#| label: stattest-ttest-ex1
ttest_ex1<-anes_des %>%
   svyttest(formula=I(EarlyVote2020=="Yes")~0,
            design=.,
            na.rm=TRUE)

> ttest_ex1

Reviewer: Opportunity to introduce broom::tidy here. Something like:

```
broom::tidy(ttest_ex1)

> broom::tidy(ttest_ex1)
# A tibble: 1 × 8
  estimate statistic  p.value parameter conf.low conf.high method                         alternative
     <dbl>     <dbl>    <dbl>     <dbl>    <dbl>     <dbl> <chr>                          <chr>      
1    0.948      226. 7.17e-77        50    0.939    0.0440 Design-based one-sample t-test two.sided  
```

Reviewer: Something is wrong here and I'm not sure if it's the package. 94.8% of people DIDN'T early vote

````
> anes_des %>%
+   filter(!is.na(EarlyVote2020)) %>%
+   group_by(EarlyVote2020) %>%
+   summarise(
+     pct=survey_prop(proportion=TRUE, vartype = "ci")
+   )
# A tibble: 2 × 4
  EarlyVote2020    pct pct_low pct_upp
  <fct>          <dbl>   <dbl>   <dbl>
1 Yes           0.0525  0.0446  0.0616
2 No            0.948   0.938   0.955 
````

```
Note that because `EarlyVote202` is a factor, we need to specify which level we are interested in.  In this case we want to isolate those that did vot early in 2020.  The results indicate that on average `r round(ttest_ex1$estimate*100,1)`% of the U.S. voting eligible population voted early in the 2020 election.  We also see that the t-statistic is `r ttest_ex1$statistic` and the p-value is `r ttest_ex1$p.value`, indicating that the average is statistically different from 0 at an $\alpha$ level of $0.05$.

> <!--Add in callout box about how to use the $ notation to help call out the different values?  Maybe indicate how this will be covered more in the reporting chapter?-->

Reviewer: This is where I think the tidy::broom will help.

### Syntax {#stattest-chi-syntax}

As with t-tests, when we do not have survey data, we may be able to use the `chisq.test()` function.  However, this function does not allow for weights or the variance structure to be accounted for with survey data. Therefore, when using survey data, we need to use one of two functions:

-  `svygofchisq()`: For goodness of fit tests
-  `svychisq()`: For tests of independence and homogeneity

The non-survey data function of `chisq.test()` requires either a single set of counts and given proportions (for goodness of fit tests), or two sets of counts for tests of independence and homogeneity.  The functions that we use with survey data require respondent level data and formulas instead of counts.  This ensures that the variances are correctly calculated.

First, the function for the goodness of fit tests is `svygofchisq()`:

> ``` {r}
> #| label: stattest-chi-gofsyntax
> #| eval: FALSE
> svygofchisq(formula,
>             p,
>             design, 
>             na.rm=TRUE,
>             ...)
> ```

Reviewer: ````suggestion
```
svygofchisq(formula,
            p,
            design, 
            na.rm=TRUE,
            ...)
```
````

The function for tests of independence and homogeneity (`svychisq()`) is similar to the goodness of fit function in that the `formula` argument is first.  However, instead of an argument for the expected proportions, the `svychisq()` function has an argument to select the statistic used for the test:

> ``` {r}
> #| label: stattest-chi-chisqsyntax
> #| eval: FALSE
> svychisq(formula,
>          design, 
>          statistic = c("F",  "Chisq", "Wald", "adjWald", 
>                        "lincom", "saddlepoint"),
>          na.rm=TRUE,
>          ...)
> ```

Reviewer: ````suggestion
```
svychisq(formula,
         design, 
         statistic = c("F",  "Chisq", "Wald", "adjWald", 
                       "lincom", "saddlepoint"),
         na.rm=TRUE,
         ...)
```
````

-  27% had a High School degree
-  29% had some college or associate's degree
-  33% had a bachelor's degree or higher

If we want to see if the weighted distribution from the ANES 2020 data matches this distribution, we could set up the hypothesis as follows:

- $H_0: p_1 = 0.11, ~ p_2 = 0.27, ~ p_3 = 0.29, ~ p_4 = 0.33$
- $H_A:$ at least one of the education levels does not match between the ANES and the ACS

To conduct this in R, we first need to adjust the education variable we have on ANES (`Education`) to have the same categories as the data from the ACS.  We do this using the {forcats} package and the `fct_collapse()` function. The education data from ANES has two levels for Bachelor's Degree or Higher (Bachelor's and Graduate), so these two categories need to be collapsed into a single category to match the ACS data.  Then we will use the `svygofchisq()` function to compare the ANES data to the ACS data:
```{r}
#| label: stattest-chi-ex1
anes_des_educ<-anes_des %>% 
  mutate(Education2=
           fct_collapse(Education,
                        "Bachelor or Higher"=c("Bachelor's",
                                               "Graduate")))

> 

Reviewer: Add a quick summary of the proportions in ANES. Something like:

```
anes_des_educ %>%
  group_by(Education2) %>%
  filter(!is.na(Education2)) %>%
  summarise(p=survey_mean())

```

chi_ex1<-anes_des_educ %>% 
  svygofchisq(formula=~Education2,
              p=c(0.11,0.27,0.29,0.33),
              design=.,
              na.rm=TRUE)

chi_ex1
```

> The output from the `svygofchisq()` indicates that at least one proportion from ANES does not match the ACS data ($\chi^2=$`r chi_ex1$statistic`; $p-value=$`r chi_ex1$p.value`. To get a better idea of the differences, we can use the `expected` output along with `survey_mean()` to create a comparison table:

Reviewer: This doesn't render correctly in HTML but does ok in PDF. 

![image](https://user-images.githubusercontent.com/8659598/235373516-943a8819-12e7-4b1b-9f78-3bccd5a82833.png)

Author: I fixed the $ issue, but it is still displaying with scientific notation.  Thoughts on how to fix that?

  group_by(TrustPeople,TrustGovernment) %>% 
  summarize(Observed=round(survey_mean(vartype = "ci"),3)) 

chi_ex2_obs %>% 
  mutate(prop=paste0(Observed," (",Observed_low,", ",
                     Observed_upp,")")) %>% 
  select(TrustGovernment,TrustPeople,prop) %>% 
  pivot_wider(names_from=TrustPeople,values_from=prop) %>% 
  gt(rowname_col = "TrustGovernment") %>% 
  tab_stubhead(label = "Trust in Government") %>% 
  tab_spanner(
    label = "Trust in People",
    columns = everything()
  )
```
Both methods produce the same output as the the `svychisq()` function does account for the survey design.  However, calculating the proportions directly from the design object means that you can also obtain the variance information.  In this case, the table output displays the survey estimate followed by the confidence intervals.  Based on the output, we can see that of those that never trust people 50.3% also never trust the government, while the proportions of never trusting the government are much lower for each of the other levels of trusting people.  

We may find it easier to look at these proportions graphically.  We can use `gglot()` and facets to provide an overview:

> <!--Need to clean this up to make it readable. Need to also add in a higher facet title to indicate that is trusting people-->

Reviewer: Perhaps a mosaic plot instead?
